File: today2/word-break.ts
/**
 * https://leetcode.com/problems/word-break/description/
 */
export { wordBreak };

function wordBreak(s: string, wordDict: string[]): boolean {
  const memo: Record<string, boolean> = {};

  function recurse(str: string): boolean {
    if (str in memo) return memo[str];
    if (str.length === 0) return true;

    for (let word of wordDict) {
      let matches = true;
      for (let w = 0; w < word.length; w++) {
        if (str[w] !== word[w]) {
          matches = false;
          break;
        }
      }
      if (matches) {
        const ans = recurse(str.slice(word.length));
        if (ans) return (memo[str] = true);
      }
    }

    return (memo[str] = false);
  }

  return recurse(s);
}

File: returned/python/today11/word_break.py
# https://leetcode.com/problems/word-break/description/
from typing import List

class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        memo = {}

        def dfs(l: int, r: int) -> bool:
            if l == len(s) and r == len(s):
                return True
            if r >= len(s):
                return False
            if (l,r) in memo:
                return memo[(l,r)]
            
            result = False
            if s[l:r+1] in wordDict:
                result = dfs(r+1, r+1)

            memo[(l,r)] = result or dfs(l, r+1)
            return memo[(l,r)]

        return dfs(0,0)
